### DISCUSSION TITLE
Strongswan with Dual band IP and both IPv4 and IPv6

### DISCUSSION THREAD

**Discussion by ssai12** [strongswan/strongswan#2196] [Q&A]
Hi Team,

We are using strongswan for encrypting our GRE tunnel between our clustered vms.
We hit a issue with one of our customer setup where they have configured dual IP like one out-bound IP (exposed to outside world) and in-bound (IP used for in-between VM communications).

There is also a docker network in-between the VMs which is where the strongswan is actually getting used for encrypting the GRE tunnel b/n docker network.

Sample response from one of the VM which shows the IP address of various **interfaces.**
# ip -br a
lo               UNKNOWN        127.0.0.1/8 ::1/128
**ens192           UP             172.18.42.115/20 - Outbound
ens224           UP             192.168.240.115/24 - Inbound
ens256           UP             192.168.241.115/24 - another inbound**
**docker0          UP             192.168.2.250/24 fd00:1234:2::1/80 fe80::1/64**
gre0@NONE        DOWN
gretap0@NONE     DOWN
erspan0@NONE     DOWN
**gre1@NONE        UNKNOWN        192.168.2.101/32 fd00:1234:2::101/128**
veth6f403ad@if94 UP             fe80::e85e:62ff:feed:3f68/64
veth9f43242@if96 UP             fe80::a0ce:67ff:fea4:72b4/64
veth874a426@if98 UP             fe80::945e:1bff:fe55:1e3c/64
veth5bb7c80@if100 UP             fe80::2cac:afff:fe83:793a/64
veth339b6b2@if102 UP             fe80::e4c8:79ff:fe9c:973e/64
vethf129d3a@if104 UP             fe80::b8d4:89ff:fe82:5d22/64
veth095f593@if106 UP             fe80::f89a:e6ff:fe99:89cd/64
vethc3ced09@if108 UP             fe80::ec18:7aff:fe0c:2cd9/64
vethf658f11@if110 UP             fe80::181f:f5ff:fe44:abf3/64
vethd1b1b30@if112 UP             fe80::88f6:a4ff:fef6:8fc/64
veth259f4cb@if114 UP             fe80::8867:c0ff:fede:88f5/64
veth824434a@if116 UP             fe80::541b:13ff:fe8a:48eb/64
veth2c97205@if118 UP             fe80::c861:93ff:fe69:1725/64
veth66bfc1a@if120 UP             fe80::104f:7dff:fe7a:ff79/64
veth8eba14c@if122 UP             fe80::8eb:1ff:fe11:c250/64
veth528541b@if124 UP             fe80::2890:92ff:fe6d:2d9/64
veth396a4c3@if126 UP             fe80::2829:f1ff:fe9c:9aea/64
veth57a257c@if128 UP             fe80::5865:faff:fe11:b35b/64
veth5908107@if130 UP             fe80::549c:f8ff:fe89:2478/64
vethb5e88e2@if132 UP             fe80::4808:47ff:fec2:e052/64
veth54cf5b9@if134 UP             fe80::ec:f4ff:fe1d:73dc/64
veth20a1cd9@if136 UP             fe80::546a:23ff:fef2:1f00/64
veth029e992@if138 UP             fe80::14c6:fdff:fe7e:2a8c/64
veth2dd7a0f@if140 UP             fe80::9082:87ff:febb:d046/64
veth3b2bd83@if142 UP             fe80::f052:2fff:fe4e:1c4f/64
veth67e3155@if144 UP             fe80::d013:8dff:fe1f:f524/64
vethfb0e360@if146 UP             fe80::dc27:7dff:fecf:a696/64
veth8d1a990@if148 UP             fe80::b429:8cff:fe78:a34b/64
veth6e049ab@if150 UP             fe80::fcf5:5bff:fe46:4c3b/64
vethac05b52@if152 UP             fe80::287d:51ff:fea3:b5fd/64
vethcc32991@if154 UP             fe80::54:edff:fee2:4e45/64
vethe105d09@if156 UP             fe80::7c43:6fff:fe60:6b32/64
vethf8bd79a@if158 UP             fe80::f415:e1ff:fedf:958b/64
vethc26ca41@if160 UP             fe80::bc40:7aff:fe7e:7532/64
veth667b868@if162 UP             fe80::5454:baff:febf:44a6/64
veth5269a65@if164 UP             fe80::80f:e1ff:fe71:6616/64
vethb088351@if166 UP             fe80::64ee:eeff:fe27:238c/64
vethd1bd780@if168 UP             fe80::48af:89ff:fe0a:ac11/64

and, below is the swanctl conf configuration
```
connections {
    dc-net {
      remote_addrs = 192.168.241.110
      local {
        auth = pubkey
        certs = client2Cert.pem
      }
      remote {
        auth = pubkey
        id = "C=CH, O=strongSwan, CN=device1"
      }
      children {
        dc-dsnet {
          mode = tunnel
          start_action = trap
          local_ts = dynamic[gre]
          remote_ts = dynamic[gre]
          esp_proposals=aes256-sha2_256
        }
      }
      proposals = aes256-sha2_256-modp1024
    }
  }

# Section defining attributes of certification authorities.
 authorities {

    # Section defining a certification authority with a unique name.
     dc1autho {

        # CA certificate belonging to the certification authority.
         cacert = strongswanCert.pem

        # Absolute path to the certificate to load.
         file = /etc/strongswan/swanctl/x509ca/

 }

}
```

# Include config snippets
include conf.d/*.conf

Issue#1
so, one node is 192.168.241.115 and another node is 192.168.241.110

so, the actual issue is, after loading the configuration using swanctl -q while initiating connection using swanctl -i --child dc-dsnet

we are seeing connection getting failed because of peer not responding. 
but, while we initial connection from one node and also connection initiation from second node the strongswan connection is getting successfully established. 

Issue#2
when we try to initiate connection from node1 it is sending the packets on 500 ports like below 
[ENC] generating IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) ]
[NET] sending packet: from 192.168.241.110[500] to 192.168.241.115[500] (336 bytes)
[IKE] retransmit 1 of request with message ID 0
[NET] sending packet: from 192.168.241.110[500] to 192.168.241.115[500] (336 bytes)
[IKE] retransmit 2 of request with message ID 0
[NET] sending packet: from 192.168.241.110[500] to 192.168.241.115[500] (336 bytes)

At the same time if we send request from another node it is sending packets on 4500 ports.

Could you please elaborate this behaviour and also any clue on why the connection is not getting established only when trying from one side.

**Comment by tobiasbrunner** [comment]
Sorry, I don't get what your question is exactly. And your topology description is also very confusing. As for problems with not getting the connection established, there are many possible reasons (firewalls, routing, config issues etc.).

**Comment by ssai12** [comment]
I have two topologies with 1*1 model 
Setup #1
IPv6 address on our Docker0 and GRE interface like below
# ip -br a
docker0          UP             192.168.1.250/24 fd00:1234:1::1/80 fe80::1/64
gre2@NONE        UNKNOWN        192.168.1.102/32 fd00:1234:1::102/128

Setup#2
Scenario with only IPv4 
# ip -br a
docker0          UP             192.168.1.250/24
gre2@NONE        UNKNOWN        192.168.1.102/32
And with below swanctl.conf

```
connections {
    dc-net {
      remote_addrs = 192.168.241.110
      local {
        auth = pubkey
        certs = client2Cert.pem
      }
      remote {
        auth = pubkey
        id = "C=CH, O=strongSwan, CN=device1"
      }
      children {
        dc-dsnet {
          mode = tunnel
          start_action = trap
          local_ts = dynamic[gre]
          remote_ts = dynamic[gre]
          esp_proposals=aes256-sha2_256
        }
      }
      proposals = aes256-sha2_256-modp1024
    }
  }

# Section defining attributes of certification authorities.
 authorities {

    # Section defining a certification authority with a unique name.
     dc1autho {

        # CA certificate belonging to the certification authority.
         cacert = strongswanCert.pem

        # Absolute path to the certificate to load.
         file = /etc/strongswan/swanctl/x509ca/

 }

}
```
With above swanctl conf initiator in Setup#1 is re-transmitting for 5 attempts and failing with peer not responding. But, strange thing is if we are initialing swanctl -i from both nodes it is successfully getting established.
And, same swanctl conf in Setup#2 is working fine.

There is a NAT and we are doing port forwarding for 500 and 4500 on both setups as well.

**Reply by tobiasbrunner** [reply]
Sounds like you have to fix your NAT/firewall/routing issue.

**Reply by ssai12** [reply]
Ok, so separating NAT/Firewall/routing issues. Is my above swanctl conf needs any adjustments or its already enough to handle IPv6?

**Reply by tobiasbrunner** [reply]
If you transport IPv6 via GRE over IPv4, possibly.

**Reply by ssai12** [reply]
Can you please help me with a example swanctl.conf on "How to transport ipv6 via GRE over IPv4"

**Reply by tobiasbrunner** [reply]
What do you mean? You already have that configured (an IPv4 SA that transports GRE), you only need to route IPv6 via that GRE tunnel.

**Reply by ssai12** [reply]
is there a way disable or prevent StrongSwan's swanctl from considering ADD_6_ADDR notifications during IKEv2 negotiation?

**Reply by tobiasbrunner** [reply]
Do you even know what these are for? If not, see [MOBIKE](https://docs.strongswan.org/docs/5.9/features/mobike.html).
